src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs

```diff
@@ -74,44 +74,123 @@ public HubConnectionContext(ConnectionContext connectionContext, TimeSpan keepAl
 
         public int? LocalPort => Features.Get<IHttpConnectionFeature>()?.LocalPort;
 
-        public virtual async Task WriteAsync(HubMessage message)
+        public virtual Task WriteAsync(HubMessage message)
         {
-            await _writeLock.WaitAsync();
+            // We were unable to get the lock so take the slow async path of waiting for the semaphore
+            if (!_writeLock.Wait(0))
+            {
+                return WriteSlowAsync(message);
+            }
+
+            // This method should never throw synchronously
+            var task = WriteCore(message);
+
+            // The write didn't complete synchronously so await completion
+            if (!task.IsCompleted)
+            {
+                return CompleteWriteAsync(task);
+            }
+
+            // The write failed so observe the exception
+            if (task.IsFaulted)
```

Comment:
I think if you use `IsCompletedSuccessfully` above then you'll get the exception observed automatically when the task is passed to `CompleteWriteAsync` and you don't need this. It is slightly slower but it is an exceptional situation so perf doesn't matter as much.